### DISCUSSION TITLE
@lit/localize static import in runtime mode

### DISCUSSION THREAD

**Discussion by jeromecambon** [lit/lit#2227] [Q&A] [answered]
Hi,

I'd like to have static import for locale files, in Runtime mode.
I mean, instead of writing:

```
export const {setLocale} = configureLocalization({
  sourceLocale,
  targetLocales,
  loadLocale: (locale) => import(`./${locale}.js`),
});
```
... I would like to have static import of all my locale files:

```
 import {templates_fr} from './fr.js'
 import {templates_es} from './es.js'
...
```

Do you think this is doable ?

Context: I'm dynamically importing a web component in my app, which does not have access to the locale files.

Thanks!

**Comment by aomarks** [comment, marked as answer]
Does https://github.com/lit/lit/issues/1661#issuecomment-814430658 sound like it would address your use case?

Could you provide a little more information about your setup? Is the component published to NPM? Does the top level app also require template localization, and is it also written with Lit?

**Reply by jeromecambon** [reply]
Thanks for your response, my issue is not related to dependencies, so your link does not help for this.
Hopefully in the meantime I've solved my issue in a quite simple way: 😃 

```
const template_es = import('./locales/es-ES.js');
const template_fr = import('./locales/fr.js');
const template_ja = import('./locales/ja.js');
const template_pt = import('./locales/pt-BR.js');
```

```
export const {setLocale} = configureLocalization({
  sourceLocale,
  targetLocales,
  loadLocale: (locale) => {
    switch (locale) {
      case 'es-ES': {
        return template_es;
      }
      case 'fr': {
        return template_fr;
      }
      case 'ja': {
        return template_ja;
      }
      case 'pt-BR': {
        return template_pt;
      }
      default: {
        return template_fr;
      }
  }}
});

```

**Reply by aomarks** [reply]
Ah, I see! A similar pattern to your example is this:

```js
export const {setLocale} = configureLocalization({
  sourceLocale,
  targetLocales,
  loadLocale: (locale) => import(`./locales/${locale}.js`)
});
```

The difference is that in your example, all locale templates will be fetched regardless of whether a user switches to that locale, whereas in this example they will only be fetched on-demand. Pre-fetching might be what you want -- e.g. if you have a very small number of localized templates and want switching to be very fast at the expensive of more total bytes fetched.

Another way to write the same thing as your example without having enumerate all locales could be this:

```js
const localizedTemplates = new Map(
  targetLocales.map((locale) => [locale, import(`/locales/${locale}.js`)])
);

export const {setLocale} = configureLocalization({
  sourceLocale,
  targetLocales,
  loadLocale: (locale) => localizedTemplates.get(locale),
});
```

And yet another variation would be to use static imports instead of dynamic imports (maybe this is the answer to your original question?). This could be useful if you need to bundle your app into a single `.js` file that includes all locales:

```js
import * as template_es from './locales/es-ES.js';
...

export const {setLocale} = configureLocalization({
  sourceLocale,
  targetLocales,
  loadLocale: (locale) => {
    switch (locale) {
      case 'es-ES': {
        return template_es;
      }
      ...
    }
  }}
});

```

**Reply by aomarks** [reply]
I'm also curious to understand a little more about why you want static imports, just to help me understand how people are using localize. Why does "dynamically importing a web component in my app, which does not have access to the locale files" restrict how your locales are loaded?

**Reply by jeromecambon** [reply]
I am in an environment where I'm loading my web component bundle js file through a REST API, then interpret it. So obviously the locale files are not available.
Anyway, thank you for your answers!

**Reply by aomarks** [reply]
We've added a section to the lit.dev localization docs that demonstrates a few approaches for loading locale modules, including the static imports pattern discussed here: https://lit.dev/docs/localization/runtime-mode/#approaches-for-loading-locale-modules

**Comment by nemethmik** [comment]
@aomarks  I have just completed my PoC application using localize, and I am totally happy with it.

**Reply by aomarks** [reply]
That doesn't seem right -- translating html-tagged literals is one of the primary use cases for Lit localize.

Can you give some more information? Are you seeing an error when you try to localize an html template?

**Reply by nemethmik** [reply]
You were right, many appologies! Here is the only situation when I had to use the *str* tagged literal:
```ts
 locStr(v:string):string {
    return msg(`A string ${v}`)
  }
```
Here I had to use *str* to work with *lit-localize extract* to avoid `error TS2324: String literal with expressions must use the str tag` 
```ts
locStr(v:string):string {
    return msg(str`A string ${v}`)
  }
```
When I first tried, I changed it to
```ts
locStr(v:string):string {
    return msg(html`A string ${v}`)
  }
```
It didn't work, of course, and I had the false conclusion that msg doesn't work with html tagged literals, but it does excellently. That was my mistake, So, I will remove my silly comments. Thank you again for the great help.

**Reply by aomarks** [reply]
No problem! Glad it's working for you, feel free to post more discussions if you have more questions.
